### ISSUE TITLE
Requirements for a built-in http client

### ISSUE THREAD

**Issue by borkdude** [borkdude/babashka#220]
Would be nice to have in addition to clj-http-lite:

- get effective url after redirection (without getting the contents itself)
- talk to a socket (Docker, see @lispyclouds's work)

**Comment by jwhitlark** [comment]
I have no idea what difficulty is involved, but I'd love to use this to talk to the Kubernetes API: https://github.com/dakrone/clj-http/blob/3.x/examples/kubernetes_pod.clj

**Comment by borkdude** [comment]
Keeping an eye on https://github.com/gnarroway/hato. This is built on JDK11's built in HTTP client and might work with GraalVM (unlike clj-http). It supports async requests and websockets. Depends on #120.

**Comment by jwhitlark** [comment]
I'll try to test it against the kubernates api later this week.

**Comment by borkdude** [comment]
First result of testing with hato on JDK11 wasn't succesful:

https://pastebin.com/nG546Dz0

**Comment by borkdude** [comment]
I made a [fork](https://github.com/borkdude/hato) where I got rid of the dynamic requires + ns-resolves. This worked!

```
$ ./bb "(require '[hato.client :as hc]) (-> (hc/get \"https://www.clojure.org\") :body (subs 0 20))"
"<!DOCTYPE html>\n<!--"
```

**Comment by lispyclouds** [comment]
@borkdude, would you consider adding https://github.com/jnr/jnr-unixsocket to bb as well? I can try looking into adding UNIX socket support as well. HTTP transport or other wise. Can possibly try make hato work over them too.

**Comment by lispyclouds** [comment]
Java unfotunately has no built in support for UNIX sockets 😕

**Comment by borkdude** [comment]
@lispyclouds I consider that to be a different issue. Welcome to create one and give some background to why you need this for scripting.

**Comment by lispyclouds** [comment]
@borkdude I was thinking since you mentioned that talking to Docker may be a thing you'd wanna do in the issue, unix socket support is necessary for it.

**Comment by borkdude** [comment]
@lispyclouds I know, but to keep things organized, it's better to put this in a different issue, with some examples why it's useful for scripting - can't keep it all in my head :).

**Comment by borkdude** [comment]
Got a build working using Github actions:

https://github.com/borkdude/babashka/actions/runs/46649842

It does take 8 minutes to build and yields a binary of nearly 55MB (17MB zipped).

Thought:

Adding clj-http-lite would result in lower compile times and maybe covers the 80% of what people want to do in a script? The rest like talking to a unix socket could be covered with clojure.java.shell + curl. It probably doesn't get any better than curl when it comes to supporting edge cases? We would not have async requests that hato supports, but that doesn't seem so common in scripting anyway.

**Comment by borkdude** [comment]
For now I'll decide to support HTTP via the following options:

https://github.com/borkdude/babashka#http

**Comment by borkdude** [comment]
Here is an avenue I want to explore. Contribs welcome. 

https://github.com/borkdude/babashka.curl

**Comment by borkdude** [comment]
Added babashka.curl to bb. I think this will suffice for most scripting duties.
